File: haskell/2021/day_12.hs
import qualified Data.Char as Char
import qualified Data.List.Split as Split
import qualified Data.Map as Map
import qualified Data.Set as Set

-- | Return the number of paths from the current node to the end.
explore :: Bool -> Map.Map String (Set.Set String) -> Set.Set String -> String -> Int
explore _ _ _ "end" = 1
explore extraTime graph visited current = sum $ do
    adj <- Set.elems $ graph Map.! current
    -- Explore all large or unvisited caves
    if all Char.isUpper adj || adj `Set.notMember` visited
        then return $ explore extraTime graph (Set.insert adj visited) adj
    -- Explore one small cave twice
    else if extraTime && adj /= "start"
        then return $ explore False graph visited adj
    else []

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_12.txt"
    let graph = Map.fromListWith Set.union $ do
        line <- lines text
        let a:b:_ = Split.splitOn "-" line
        (start, end) <- [(a, b), (b, a)]
        return (start, Set.singleton end)
    let visited = Set.singleton "start"
    putStrLn $ "Part 1: " ++ show (explore False graph visited "start")
    putStrLn $ "Part 1: " ++ show (explore True graph visited "start")

File: rust/2021/src/bin/day_12.rs
use std::collections::{HashMap, HashSet};
use std::fs;

type Graph<'a> = HashMap<&'a str, Vec<&'a str>>;

fn explore<'a>(
    graph: &Graph<'a>,
    visited: &mut HashSet<&'a str>,
    current: &'a str,
    extra_time: bool,
) -> usize {
    if current == "end" {
        return 1;
    }
    let mut paths = 0;
    for adj in &graph[current] {
        if adj.bytes().all(|b| b.is_ascii_uppercase()) || !visited.contains(adj) {
            visited.insert(adj);
            paths += explore(graph, visited, adj, extra_time);
            visited.remove(adj);
        } else if extra_time && *adj != "start" {
            paths += explore(graph, visited, adj, false);
        }
    }
    paths
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_12.txt").unwrap();
    let mut graph: Graph = HashMap::new();
    for line in input.lines() {
        let (a, b) = line.split_once('-').unwrap();
        for (start, end) in [(a, b), (b, a)] {
            graph.entry(start).or_default().push(end);
        }
    }
    let mut visited = HashSet::from(["start"]);
    println!("Part 1: {}", explore(&graph, &mut visited, "start", false));
    println!("Part 2: {}", explore(&graph, &mut visited, "start", true));
}
